### PR TITLE
Migrate ProtobufExtract task to lazy configuration api

### PR DESCRIPTION
**Pull request by rougsig**
Background:
Fix for https://github.com/google/protobuf-gradle-plugin/issues/537

Changes:
Migrate ProtobufExtract task to [gradle lazy configuration api](https://docs.gradle.org/current/userguide/lazy_configuration.html#lazy_properties).
Mark necessary properties with `Input`, `Internal`, `OutputDirectory`, etc annotations

Test plan:
Green pipelines, production code logic was not changed. Check build deprecations with `--warning-mode=all`

IMPORTANT NOTES:
`ProtobufExtract` api was changed. Its breaking change. In reason of `ProtobufExtract` is internal use, there shouldn't be any problems.

### CODE REVIEW

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy [hunk 1]
```diff
@@ -58,80 +58,92 @@ import javax.inject.Inject
...
-  public ConfigurableFileCollection getInputFiles() {
-    return inputFiles
-  }
+  public abstract ConfigurableFileCollection getInputFiles()
 
-  /**
-   * Inputs for this task containing only proto files, which is enough for up-to-date checks.
-   * Add inputs to inputFiles. Uses relative path sensitivity as directory layout changes impact output.
-   */
+  // Inputs for this task containing only proto files, which is enough for up-to-date checks.
```
Comment (ejona86) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy:76]: Why this change?

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy [hunk 2]
```diff
@@ -58,80 +58,92 @@ import javax.inject.Inject
...
-
-  @Internal
-  CopyActionFacade getCopyActionFacade() {
-    return copyActionFacade
-  }
-
-  @Internal
-  ArchiveActionFacade getArchiveActionFacade() {
-    return archiveActionFacade
+  public FileTree getInputProtoFiles() {
```
Comment (ejona86) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy:80]: Why this change?

**Review by ejona86 — COMMENTED** [review]
Were you able to reproduce #537? I've not been able to reproduce it and the released version of the task does have the property annotated with `@OutputDirectory`. My only idea for it is something doesn't like the `protected` access, but that still doesn't explain why I can't reproduce.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 2 above)
Comment (rougsig) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy:80]: IMO `getInputProtoFiles` better name than `getFilteredProtosFromInputs`.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy [hunk 3]
```diff
@@ -58,80 +58,92 @@ import javax.inject.Inject
...
+  @Inject
+  protected abstract ObjectFactory getObjectFactory()
 
-  @OutputDirectory
-  protected File getDestDir() {
-    return destDir
+  @Inject
+  protected abstract ProviderFactory getProviderFactory()
+
+  private FileCollection instantiateFilteredProtos() {
```
[the remark this answers is not in the document]
Comment (rougsig) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: oops my bad. Didn't pay enough attention.

**Comment by rougsig** [comment]
> Were you able to reproduce https://github.com/google/protobuf-gradle-plugin/issues/537? I've not been able to reproduce it and the released version of the task does have the property annotated with @OutputDirectory. My only idea for it is something doesn't like the protected access, but that still doesn't explain why I can't reproduce.

Reproduced on gradle 7.3.1. This looks like a Gradle bug, but I can't be sure, couldn't find an issue or something like that. In the new gradle version, not reproducible. To be fair, this change is more for newer gradle api, like provider. I think this bug it not necessary to fix, in reason of hard to reproduce.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 2 above)
Comment (ejona86) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy:86]: Okay, that's fine.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: Could you revert the whitespace changes for this function? I see how the new formatting looks, but the old formatting wasn't wrong.

[the remark this answers is not in the document]
Comment (rougsig) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: Done

**Review by ejona86 — APPROVED** [review]

**Review by YifeiZhuang — APPROVED** [review]
